### Automated preprocessing with scikit-learn pipelines

In [2]:
from warnings import filterwarnings
filterwarnings('ignore')

### Step 1: Read the dataset

In [3]:
import pandas as pd

df = pd.read_csv("Cars93.csv", na_values=['','NA'], keep_default_na=False)
df.head()

id Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0   1        Acura  Integra    Small       12.9   15.9       18.8        25   
1   2        Acura   Legend  Midsize       29.2   33.9       38.7        18   
2   3         Audi       90  Compact       25.9   29.1       32.3        20   
3   4         Audi      100  Midsize       30.8   37.7       44.6        19   
4   5          BMW     535i  Midsize       23.7   30.0       36.2        22   

   MPG.highway             AirBags  ... Passengers Length  Wheelbase  Width  \
0           31                None  ...          5    177        102     68   
1           25  Driver & Passenger  ...          5    195        115     71   
2           26         Driver only  ...          5    180        102     67   
3           26                 NaN  ...          6    193        106     70   
4           30         Driver only  ...          4    186        109     69   

   Turn.circle  Rear.seat.room Luggage.room  Weight   Origin           Make  
0           37            26.5         11.0    2705  non-USA  Acura Integra  
1           38            30.0         15.0    3560  non-USA   Acura Legend  
2           37            28.0         14.0    3375  non-USA        Audi 90  
3           37            31.0         17.0    3405  non-USA       Audi 100  
4           39            27.0         13.0    3640  non-USA       BMW 535i  

[5 rows x 28 columns]

### Step 2: Perform Basic data quality checks

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93 entries, 0 to 92
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  93 non-null     int64  
 1   Manufacturer        93 non-null     object 
 2   Model               93 non-null     object 
 3   Type                93 non-null     object 
 4   Min.Price           93 non-null     float64
 5   Price               93 non-null     float64
 6   Max.Price           93 non-null     float64
 7   MPG.city            93 non-null     int64  
 8   MPG.highway         93 non-null     int64  
 9   AirBags             89 non-null     object 
 10  DriveTrain          93 non-null     object 
 11  Cylinders           93 non-null     object 
 12  EngineSize          93 non-null     float64
 13  Horsepower          93 non-null     int64  
 14  RPM                 93 non-null     int64  
 15  Rev.per.mile        93 non-null     int64  
 16  Man.trans.

In [5]:
m = df.isna().sum()
m

id                     0
Manufacturer           0
Model                  0
Type                   0
Min.Price              0
Price                  0
Max.Price              0
MPG.city               0
MPG.highway            0
AirBags                4
DriveTrain             0
Cylinders              0
EngineSize             0
Horsepower             0
RPM                    0
Rev.per.mile           0
Man.trans.avail        0
Fuel.tank.capacity     0
Passengers             0
Length                 0
Wheelbase              0
Width                  0
Turn.circle            0
Rear.seat.room         2
Luggage.room          11
Weight                 0
Origin                 0
Make                   0
dtype: int64

In [6]:
m[m>0]

AirBags            4
Rear.seat.room     2
Luggage.room      11
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

### Step 3 : Seperate X and Y(Weight)

Weight ~ remaining feature

In [8]:
X = df.drop(columns=['id', 'Weight'])
Y = df[['Weight']]


In [9]:
X.head()

Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0        Acura  Integra    Small       12.9   15.9       18.8        25   
1        Acura   Legend  Midsize       29.2   33.9       38.7        18   
2         Audi       90  Compact       25.9   29.1       32.3        20   
3         Audi      100  Midsize       30.8   37.7       44.6        19   
4          BMW     535i  Midsize       23.7   30.0       36.2        22   

   MPG.highway             AirBags DriveTrain  ... Fuel.tank.capacity  \
0           31                None      Front  ...               13.2   
1           25  Driver & Passenger      Front  ...               18.0   
2           26         Driver only      Front  ...               16.9   
3           26                 NaN      Front  ...               21.1   
4           30         Driver only       Rear  ...               21.1   

   Passengers  Length  Wheelbase  Width Turn.circle  Rear.seat.room  \
0           5     177        102     68          37            26.5   
1           5     195        115     71          38            30.0   
2           5     180        102     67          37            28.0   
3           6     193        106     70          37            31.0   
4           4     186        109     69          39            27.0   

   Luggage.room   Origin           Make  
0          11.0  non-USA  Acura Integra  
1          15.0  non-USA   Acura Legend  
2          14.0  non-USA        Audi 90  
3          17.0  non-USA       Audi 100  
4          13.0  non-USA       BMW 535i  

[5 rows x 26 columns]

In [10]:
Y.head()

Weight
0    2705
1    3560
2    3375
3    3405
4    3640

### Step 4 : Preprocessing on X

In [11]:
cat = list(X.columns[X.dtypes=='object'])
con = list(X.columns[X.dtypes!='object'])

In [12]:
cat

['Manufacturer',
 'Model',
 'Type',
 'AirBags',
 'DriveTrain',
 'Cylinders',
 'Man.trans.avail',
 'Origin',
 'Make']

In [13]:
con

['Min.Price',
 'Price',
 'Max.Price',
 'MPG.city',
 'MPG.highway',
 'EngineSize',
 'Horsepower',
 'RPM',
 'Rev.per.mile',
 'Fuel.tank.capacity',
 'Passengers',
 'Length',
 'Wheelbase',
 'Width',
 'Turn.circle',
 'Rear.seat.room',
 'Luggage.room']

In [14]:
from sklearn.pipeline import Pipeline # to automate the preprocessing
from sklearn.impute import SimpleImputer # Relacing missing values
from sklearn.preprocessing import StandardScaler , OneHotEncoder# Preprocessing
from sklearn.compose import ColumnTransformer # combine pipelines

In [15]:
num_pipe = Pipeline(steps=[('imputer', SimpleImputer(strategy='mean')), ('scaler', StandardScaler())])

In [16]:
cat_pipe = Pipeline(steps=[('imputer', SimpleImputer(strategy='most_frequent')),
                           ('ohe', OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

In [17]:
pre = ColumnTransformer([('num', num_pipe, con), ('cat', cat_pipe, cat)]).set_output(transform='pandas')

In [18]:
X_pre = pre.fit_transform(X)
X_pre

num__Min.Price  num__Price  num__Max.Price  num__MPG.city  \
0        -0.485787   -0.375720       -0.282465       0.471312   
1         1.388017    1.497844        1.531409      -0.781032   
2         1.008658    0.998227        0.948052      -0.423219   
3         1.571949    1.893374        2.069191      -0.602126   
4         0.755752    1.091905        1.303535      -0.065407   
..             ...         ...             ...            ...   
88       -0.060445    0.019810        0.073018      -0.959938   
89        0.054512    0.051036        0.045673      -0.244313   
90        0.663786    0.394523        0.164167      -0.781032   
91        0.537333    0.332071        0.145937      -0.244313   
92        0.882205    0.748418        0.601684      -0.423219   

    num__MPG.highway  num__EngineSize  num__Horsepower  num__RPM  \
0           0.360925        -0.841022        -0.073484  1.717489   
1          -0.770514         0.515869         1.078322  0.369586   
2          -0.581941         0.128186         0.540813  0.369586   
3          -0.581941         0.128186         0.540813  0.369586   
4           0.172352         0.806631         1.231897  0.706562   
..               ...              ...              ...       ...   
88         -1.524806        -0.162577        -0.668585 -1.315292   
89          0.172352        -0.647181        -0.188665  0.875050   
90         -0.770514         0.128186         0.655993  0.875050   
91         -0.204794        -0.356418        -0.572601  0.201098   
92         -0.204794        -0.259498         0.464025  1.549001   

    num__Rev.per.mile  num__Fuel.tank.capacity  ...  cat__Make_Toyota Camry  \
0            1.129530                -1.062184  ...                     0.0   
1            0.005661                 0.409445  ...                     0.0   
2           -0.105713                 0.072197  ...                     0.0   
3            0.410659                 1.359872  ...                     0.0   
4            0.430909                 1.359872  ...                     0.0   
..                ...                      ...  ...                     ...   
88           1.180155                 1.359872  ...                     0.0   
89           0.714407                 0.562740  ...                     0.0   
90           0.106911                 0.562740  ...                     0.0   
91          -0.237337                -0.265051  ...                     0.0   
92          -0.044963                 0.808011  ...                     0.0   

    cat__Make_Toyota Celica  cat__Make_Toyota Previa  cat__Make_Toyota Tercel  \
0                       0.0                      0.0                      0.0   
1                       0.0                      0.0                      0.0   
2                       0.0                      0.0                      0.0   
3                       0.0                      0.0                      0.0   
4                       0.0                      0.0                      0.0   
..                      ...                      ...                      ...   
88                      0.0                      0.0                      0.0   
89                      0.0                      0.0                      0.0   
90                      0.0                      0.0                      0.0   
91                      0.0                      0.0                      0.0   
92                      0.0                      0.0                      0.0   

    cat__Make_Volkswagen Corrado  cat__Make_Volkswagen Eurovan  \
0                            0.0                           0.0   
1                            0.0                           0.0   
2                            0.0                           0.0   
3                            0.0                           0.0   
4                            0.0                           0.0   
..                           ...                           ...   
88                    

In [19]:
pre

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('imputer', SimpleImputer()),
                                                 ('scaler', StandardScaler())]),
                                 ['Min.Price', 'Price', 'Max.Price', 'MPG.city',
                                  'MPG.highway', 'EngineSize', 'Horsepower',
                                  'RPM', 'Rev.per.mile', 'Fuel.tank.capacity',
                                  'Passengers', 'Length', 'Wheelbase', 'Width',
                                  'Turn.circle', 'Rear.seat.room',
                                  'Luggage.room']),
                                ('cat',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ohe',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 ['Manufacturer', 'Model', 'Type', 'AirBags',
                                  'DriveTrain', 'Cylinders', 'Man.trans.avail',
                                  'Origin', 'Make'])])

### Step 5 : Apply train test split

In [20]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(X_pre, Y, test_size=0.2, random_state=21)

In [21]:
xtrain.head()

num__Min.Price  num__Price  num__Max.Price  num__MPG.city  \
57        1.365026    1.289670        1.185041      -0.423219   
31       -1.003095   -0.979424       -0.911397       0.113500   
62        0.606307    0.685966        0.729294      -0.781032   
29        0.043016   -0.021825       -0.063707      -0.423219   
51        1.985795    1.726835        1.449374      -0.781032   

    num__MPG.highway  num__EngineSize  num__Horsepower  num__RPM  \
57         -0.016221        -0.356418        -0.265452 -0.304365   
31          0.172352        -0.841022        -0.323043  2.054464   
62         -0.959087         0.322027         1.116716  1.212025   
29         -0.204794         0.806631         1.347077  0.875050   
51         -0.581941         1.872759         1.270290 -1.146804   

    num__Rev.per.mile  num__Fuel.tank.capacity  ...  cat__Make_Toyota Camry  \
57           0.187910                -0.663618  ...                     0.0   
31           0.157535                -1.062184  ...                     0.0   
62          -0.247462                 0.716035  ...                     0.0   
29          -0.713210                 0.409445  ...                     0.0   
51          -0.996708                 1.022624  ...                     0.0   

    cat__Make_Toyota Celica  cat__Make_Toyota Previa  cat__Make_Toyota Tercel  \
57                      0.0                      0.0                      0.0   
31                      0.0                      0.0                      0.0   
62                      0.0                      0.0                      0.0   
29                      0.0                      0.0                      0.0   
51                      0.0                      0.0                      0.0   

    cat__Make_Volkswagen Corrado  cat__Make_Volkswagen Eurovan  \
57                           0.0                           0.0   
31                           0.0                           0.0   
62                           0.0                           0.0   
29                           0.0                           0.0   
51                           0.0                           0.0   

    cat__Make_Volkswagen Fox  cat__Make_Volkswagen Passat  \
57                       0.0                          0.0   
31                       0.0                          0.0   
62                       0.0                          0.0   
29                       0.0                          0.0   
51                       0.0                          0.0   

    cat__Make_Volvo 240  cat__Make_Volvo 850  
57                  0.0                  0.0  
31                  0.0                  0.0  
62                  0.0                  0.0  
29                  0.0                  0.0  
51                  0.0                  0.0  

[5 rows x 257 columns]

In [22]:
ytrain.head()

Weight
57    2920
31    2530
62    3730
29    3490
51    4055

In [23]:
xtest.head()

num__Min.Price  num__Price  num__Max.Price  num__MPG.city  \
23       -1.003095   -0.854520       -0.701753       0.113500   
86        0.203957    0.332071        0.428500      -0.781032   
91        0.537333    0.332071        0.145937      -0.244313   
21        1.422504    1.039862        0.692834      -0.423219   
17        0.100495   -0.073868       -0.209546      -0.959938   

    num__MPG.highway  num__EngineSize  num__Horsepower  num__RPM  \
23         -0.016221        -0.453339        -0.975733 -0.809828   
86         -1.336233        -0.259498        -0.111878 -0.472853   
91         -0.204794        -0.356418        -0.572601  0.201098   
21         -0.581941         0.612789         0.060893 -0.809828   
17         -0.581941         2.260442         0.502419 -1.820755   

    num__Rev.per.mile  num__Fuel.tank.capacity  ...  cat__Make_Toyota Camry  \
23           0.532158                -0.816912  ...                     0.0   
86           0.370159                 0.961306  ...                     0.0   
91          -0.237337                -0.265051  ...                     0.0   
21          -1.108083                -0.203734  ...                     0.0   
17          -1.988953                 1.942392  ...                     0.0   

    cat__Make_Toyota Celica  cat__Make_Toyota Previa  cat__Make_Toyota Tercel  \
23                      0.0                      0.0                      0.0   
86                      0.0                      1.0                      0.0   
91                      0.0                      0.0                      0.0   
21                      0.0                      0.0                      0.0   
17                      0.0                      0.0                      0.0   

    cat__Make_Volkswagen Corrado  cat__Make_Volkswagen Eurovan  \
23                           0.0                           0.0   
86                           0.0                           0.0   
91                           0.0                           0.0   
21                           0.0                           0.0   
17                           0.0                           0.0   

    cat__Make_Volkswagen Fox  cat__Make_Volkswagen Passat  \
23                       0.0                          0.0   
86                       0.0                          0.0   
91                       0.0                          0.0   
21                       0.0                          0.0   
17                       0.0                          0.0   

    cat__Make_Volvo 240  cat__Make_Volvo 850  
23                  0.0                  0.0  
86                  0.0                  0.0  
91                  1.0                  0.0  
21                  0.0                  0.0  
17                  0.0                  0.0  

[5 rows x 257 columns]

In [24]:
ytest.head()

Weight
23    2670
86    3785
91    2985
21    3570
17    3910

In [25]:
xtrain.shape

(74, 257)

In [26]:
xtest.shape

(19, 257)

### Preprocessing done

### Step 6: Build the model

In [27]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(xtrain, ytrain)

LinearRegression()

### Step 7 : Evaluate the model

In [28]:
model.score(xtrain, ytrain)

1.0

In [29]:
model.score(xtest, ytest)

0.9331364483002006

In [30]:
ypred_train = model.predict(xtrain)
ypred_test = model.predict(xtest)

In [31]:
ypred_train[0:5]

array([[2920.],
       [2530.],
       [3730.],
       [3490.],
       [4055.]])

In [32]:
ytrain.head()

Weight
57    2920
31    2530
62    3730
29    3490
51    4055

In [33]:
ypred_test[0:5]

array([[2573.41617573],
       [3616.87436046],
       [3040.82463366],
       [3333.01612516],
       [3969.89471087]])

In [34]:
ytest.head()

Weight
23    2670
86    3785
91    2985
21    3570
17    3910

In [35]:
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error,mean_absolute_percentage_error, r2_score

In [36]:
def evaluate_model(model, x, y):
    # predict values
    ypred = model.predict(x)

    # evaluate metrics
    mse = mean_squared_error(y, ypred)
    rmse = mse**(1/2)
    mae = mean_absolute_error(y, ypred)
    mape = mean_absolute_percentage_error(y, ypred)
    r2 = r2_score(y, ypred)

    # print all the metrics
    print(f'Mean Squared Error : {mse:.2f}')
    print(f'Root Mean Squared Error : {rmse:.2f}')
    print(f'Mean Absolute Error : {mae:.2f}')
    print(f'Mean Absolute Percentage Error : {mape:.2f}')
    print(f'R2 score: {r2:.4f}')

In [37]:
evaluate_model(model, xtrain, ytrain)

Mean Squared Error : 0.00
Root Mean Squared Error : 0.00
Mean Absolute Error : 0.00
Mean Absolute Percentage Error : 0.00
R2 score: 1.0000


In [38]:
evaluate_model(model, xtest, ytest)

Mean Squared Error : 24251.21
Root Mean Squared Error : 155.73
Mean Absolute Error : 115.72
Mean Absolute Percentage Error : 0.04
R2 score: 0.9331


### AS r2 score in test is 0.9331 > 0.8  hence it is good model
### Step 8 : Predict out of sample data 

In [39]:
xnew = pd.read_csv("sample.csv", na_values=['','NA'], keep_default_na=False)
xnew

Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0         Audi      100  Midsize       30.8   37.7       44.6        19   
1      Pontiac  Sunbird  Compact        9.4   11.1       12.8        23   
2    Chevrolet   Lumina  Midsize       13.4   15.9       18.4        21   
3        Mazda     RX-7   Sporty       32.5   32.5       32.5        17   
4   Volkswagen      Fox    Small        8.7    9.1        9.5        25   

   MPG.highway      AirBags DriveTrain  ... Fuel.tank.capacity  Passengers  \
0           26          NaN      Front  ...               15.0           6   
1           31         None      Front  ...               15.2           5   
2           29         None      Front  ...               16.5           6   
3           25  Driver only       Rear  ...               20.0           2   
4           33         None      Front  ...               12.4           4   

   Length  Wheelbase  Width Turn.circle  Rear.seat.room  Luggage.room  \
0     190        106     65          37            31.0          17.0   
1     181        101     66          39            25.0          13.0   
2     198        108     71          40            28.5          16.0   
3     169         96     69          37             NaN           NaN   
4     163         93     63          34            26.0          10.0   

    Origin              Make  
0  non-USA          Audi 100  
1      USA   Pontiac Sunbird  
2      USA  Chevrolet Lumina  
3  non-USA        Mazda RX-7  
4  non-USA    Volkswagen Fox  

[5 rows x 26 columns]

In [40]:
pre


ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('imputer', SimpleImputer()),
                                                 ('scaler', StandardScaler())]),
                                 ['Min.Price', 'Price', 'Max.Price', 'MPG.city',
                                  'MPG.highway', 'EngineSize', 'Horsepower',
                                  'RPM', 'Rev.per.mile', 'Fuel.tank.capacity',
                                  'Passengers', 'Length', 'Wheelbase', 'Width',
                                  'Turn.circle', 'Rear.seat.room',
                                  'Luggage.room']),
                                ('cat',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ohe',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 ['Manufacturer', 'Model', 'Type', 'AirBags',
                                  'DriveTrain', 'Cylinders', 'Man.trans.avail',
                                  'Origin', 'Make'])])

In [41]:
xnew.shape

(5, 26)

In [42]:
xnew_pre = pre.transform(xnew)
xnew_pre

num__Min.Price  num__Price  num__Max.Price  num__MPG.city  \
0        1.571949    1.893374        2.069191      -0.602126   
1       -0.888138   -0.875337       -0.829362       0.113500   
2       -0.428309   -0.375720       -0.318925      -0.244313   
3        1.767376    1.352122        0.966282      -0.959938   
4       -0.968608   -1.083511       -1.130155       0.471312   

   num__MPG.highway  num__EngineSize  num__Horsepower  num__RPM  \
0         -0.581941         0.128186         0.540813  0.369586   
1          0.360925        -0.647181        -0.649388 -0.135877   
2         -0.016221        -0.453339        -0.649388 -0.135877   
3         -0.770514        -1.325626         2.134145  2.054464   
4          0.738071        -0.841022        -1.206095  0.369586   

   num__Rev.per.mile  num__Fuel.tank.capacity  ...  cat__Make_Toyota Camry  \
0           0.410659                -0.510323  ...                     0.0   
1           0.673908                -0.449005  ...                     0.0   
2           0.532158                -0.050439  ...                     0.0   
3          -0.014589                 1.022624  ...                     0.0   
4           0.441034                -1.307455  ...                     0.0   

   cat__Make_Toyota Celica  cat__Make_Toyota Previa  cat__Make_Toyota Tercel  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   cat__Make_Volkswagen Corrado  cat__Make_Volkswagen Eurovan  \
0                           0.0                           0.0   
1                           0.0                           0.0   
2                           0.0                           0.0   
3                           0.0                           0.0   
4                           0.0                           0.0   

   cat__Make_Volkswagen Fox  cat__Make_Volkswagen Passat  cat__Make_Volvo 240  \
0                       0.0                          0.0                  0.0   
1                       0.0                          0.0                  0.0   
2                       0.0                          0.0                  0.0   
3                       0.0                          0.0                  0.0   
4                       1.0                          0.0                  0.0   

   cat__Make_Volvo 850  
0                  0.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  

[5 rows x 257 columns]

In [43]:
xnew_pre.shape

(5, 257)

In [44]:
preds = model.predict(xnew_pre)
preds

array([[3334.22246823],
       [2575.        ],
       [3195.        ],
       [2895.        ],
       [2240.        ]])

In [46]:
df_final = xnew[['Manufacturer','Model']]
df_final

Manufacturer    Model
0         Audi      100
1      Pontiac  Sunbird
2    Chevrolet   Lumina
3        Mazda     RX-7
4   Volkswagen      Fox

In [47]:
df_final['Weight_pred'] = preds


In [48]:
df_final

Manufacturer    Model  Weight_pred
0         Audi      100  3334.222468
1      Pontiac  Sunbird  2575.000000
2    Chevrolet   Lumina  3195.000000
3        Mazda     RX-7  2895.000000
4   Volkswagen      Fox  2240.000000